packages/libs/error-stack/src/report.rs

```diff
@@ -205,13 +234,94 @@ impl<C> Report<C> {
+        #[allow(unused_mut)]
+        let mut source: Box<[Frame]> = Box::new([]);
+
+        #[cfg(all(nightly, feature = "std"))]
+        if let Some(backtrace) = backtrace {
+            let frame = Frame::from_attachment(backtrace, Location::caller(), source);
+            source = Box::new([frame]);
+        }
+
+        #[cfg(all(nightly, feature = "spantrace"))]
+        if let Some(span_trace) = span_trace {
+            let frame = Frame::from_attachment(span_trace, Location::caller(), source);
+            source = Box::new([frame]);
+        }
+
+        let frame = Frame::from_context(context, Location::caller(), source);
+
+        Self::from_frame(frame)
+    }
+
+    /// Add an existing [`Report`] as an additional source of this report.
+    ///
+    /// Should report consist of multiple frames ([`add_source()`] was called on it before),
+    /// then this merges both.
+    ///
+    /// ```rust
+    /// use std::{
+    ///     fmt::{Display, Formatter},
+    ///     path::Path,
+    /// };
+    ///
+    /// use error_stack::{Context, Report};
+    /// # #[cfg(all(not(miri), feature = "std"))]
+    /// use error_stack::{IntoReport, ResultExt};
+    ///
+    /// #[derive(Debug)]
+    /// struct IoError;
+    ///
+    /// impl Display for IoError {
+    ///     # fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+    ///     #     const _: &str = stringify!(
+    ///             ...
+    ///     #     );
+    ///     #     f.write_str("Io Error")
+    ///     # }
+    /// }
+    ///
+    /// # impl Context for IoError {}
+    ///
+    /// # #[allow(unused_variables)]
+    /// fn read_config(path: impl AsRef<Path>) -> Result<String, Report<IoError>> {
+    ///     # #[cfg(any(miri, not(feature = "std")))]
+    ///     # return Err(error_stack::report!(IoError).attach_printable("Not supported"));
+    ///     # #[cfg(all(not(miri), feature = "std"))]
+    ///     std::fs::read_to_string(path.as_ref())
+    ///         .into_report()
+    ///         .change_context(IoError)
+    /// }
+    ///
+    /// let mut error1 = read_config("config.txt").unwrap_err();
+    /// let error2 = read_config("config2.txt").unwrap_err();
+    /// let mut error3 = read_config("config3.txt").unwrap_err();
+    ///
+    /// error1.add_source(error2);
+    /// error3.add_source(error1);
+    ///
+    /// # #[allow(unused_mut)]
+    /// # let mut count = 6;
+    /// # #[cfg(all(nightly, feature = "std"))]
+    /// # {
+    /// #   count += 3;
+    /// # }
+    /// # #[cfg(feature = "spantrace")]
+    /// # {
+    /// #   count += 3;
+    /// # }
+    /// # assert_eq!(error3.frames().count(), count);
+    ///
+    /// // ^ This is equivalent to:
+    /// // error3.add_source(error1);
+    /// // error3.add_source(error2);
+    /// ```
+    ///
+    /// [`add_source()`]: Self::add_source
+    pub fn add_source<T>(&mut self, mut report: Report<T>) {
+        self.frames.append(&mut report.frames);
```

Comment:
This is not adding a `source` vertically but adding a `frame` horizontally. I don't see a reason for duplicating the `extend` implementation. We however could keep it and make it non-`pub`.

Reply:
that makes sense, I just saw it as a convenience, as otherwise, one might write:

```rust
use std::iter::once;

report.extend(once(other_report));
```

Reply:
That's a valid point. Do we want to rename it to `extend_one`? Don't have a real opinion on the name, but `source` is misleading.

Reply:
I definitely agree, I think `extend_one` is a good name!